Iceberg: Type names for Avro union structs

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

PR updates Avro union translation such that filelds take the name of their type rather than `union_opt_X` as before.

The type-to-name translation works as described in https://avro.apache.org/docs/1.12.0/specification/#json-encoding

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* (Iceberg) Struct fields corresponding to Avro union branches are given names matching their type rather than `union_opt_X`.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
ice/avro: Use JSON encoded type label as union struct field

See https://avro.apache.org/docs/1.12.0/specification/#json-encoding

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
@nvartolomei @rockwotj  - enum discussion notwithstanding, is this what we would want for complex  types? Could get unwieldy w/ lots of nesting, but in practice maybe that's a non-issue?

rockwotj commented on src/v/iceberg/conversion/schema_avro.cc:
```diff
@@ -209,8 +222,8 @@ inner_field_type_from_avro(const avro::NodePtr& node, state& state) {
-                    // <union_name>_<leaf_idx>
-                    fmt::format("union_opt_{}", i),
+                    // <field_type>
+                    fmt::format("{}", field.value()),
```
Yeah, we're following the JSON encoding here: https://avro.apache.org/docs/1.11.1/specification/#json-encoding

We need to test all the complex types with this:

https://github.com/redpanda-data/avro/blob/8dccd0d0060ad6abd773207bee91d3c70aaaf764/lang/c/src/value-json.c#L337-L376

oleiman replied:
I don't think the avro spec is sufficient here. For example, two maps of different k/v type both encode to "map", and record encoding seems to rely on the record having a "name", which is lost in iceberg translation, no?

rockwotj replied:
> which is lost in iceberg translation, no?

How is it lost in iceberg translation? We're doing conversion here so we should have all the info we need.

rockwotj replied:
>. For example, two maps of different k/v type both encode to "map",

Let me see what other avro libraries do here

rockwotj replied:
https://github.com/zolyfarkas/avro/blob/dae9b44dc61a475f8c9d1164646bc3547fcb97ee/lang/java/avro/src/main/java/org/apache/avro/io/JsonEncoder.java#L299

oleiman replied:
> How is it lost in iceberg translation? 

ah, i forgot how avro works viz that record types must be bound to a name. i was thinking of the field name one level up.

oleiman replied:
> JsonEncoder.java

this reads to me like you just get "map", unqualified. maybe it's the case that a union with two map branches is illegal then. will check.

rockwotj replied:
This seems to be adding key and value name: https://github.com/hamba/avro/blob/main/codec_union.go#L561

oleiman replied:
ya only one of each non-named type per schema, per the [spec](https://avro.apache.org/docs/1.11.1/specification/#unions):

> Unions may not contain more than one schema with the same type, except for the named types record, fixed and enum. For example, unions containing two array types or two map types are not permitted, but two types with different names are permitted.

> this seems to be adding key and value name

I could go either way. seems a bit overkill and deviates from the reference implementation, right?

rockwotj replied:
yeah let's just do the spec

oleiman committed:
ice/avro: Introduce union_branch_label(Node, field_type)

Converts a field type to a string based on the union branch JSON encoding
rules laid out in https://avro.apache.org/docs/1.12.0/specification/#json-encoding

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/avro: Fix a couple of typos in schema_avro.h

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3404203c-553d-49dc-99d7-f6ff6df0ad0f-->
<!--TEST RESULTS-->
<!--7b896b45-43b8-4758-ba42-45d941d008eb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74585">build#74585</a></summary>


| test_class                       | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                              |
|:---------------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest            | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74585#019a0589-1da2-4800-859f-e14f9bf2c493 | FLAKY         | 19/21    | upstream reliability is '98.54545454545455'. current run reliability is '90.47619047619048'. drift is 8.06926 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations  |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss           | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74585#019a0589-1da6-48e4-9274-4b3174ebe768 | FLAKY         | 19/21    | upstream reliability is '92.15219976218788'. current run reliability is '90.47619047619048'. drift is 1.67601 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74641">build#74641</a></summary>


| test_class                 | test_method            | test_arguments                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                  |
|:---------------------------|:-----------------------|:-------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTest          | compaction_stress_test | {"cleanup_policy": "compact,delete", "key_set_cardinality": 100, "storage_compaction_key_map_memory_kb": 10} | integration | https://buildkite.com/redpanda/redpanda/builds/74641#019a08e5-09a2-4ccd-9e1c-43ac6dad8c29 | FLAKY         | 20/21    | upstream reliability is '99.59839357429718'. current run reliability is '95.23809523809523'. drift is 4.3603 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTest&test_method=compaction_stress_test        |
| ShadowLinkingRandomOpsTest | test_node_operations   | {"failures": true}                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74641#019a08e5-0998-4b0c-b58b-8ab01d113166 | FLAKY         | 2/21     | upstream reliability is '32.58426966292135'. current run reliability is '9.523809523809524'. drift is 23.06046 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<!--/body-->

rockwotj commented on src/v/iceberg/conversion/tests/iceberg_avro_tests.cc:
```diff
@@ -1269,3 +1269,76 @@ INSTANTIATE_TEST_SUITE_P(
+}
+
+using btc = branch_label_test_case;
+INSTANTIATE_TEST_SUITE_P(
```
What about logical types?

oleiman replied:
hm good call out

oleiman replied:
> logical types

feels a bit fancy to encode here. logical type doesn't affect the rules for union branches nor does the ref impl do this (afaict). did you see something different in the wild?

rockwotj replied:
No I haven't seen anything, I didn't study the reference impl deeply.

oleiman replied:
I mean me either. So much Java. I'm fairly certain that what we have here is at least unambiguous. @nvartolomei any thoughts?

nvartolomei replied:
related discussions
* https://github.com/risingwavelabs/risingwave/pull/17485#discussion_r1668012859
* https://github.com/risingwavelabs/risingwave/issues/17616

Hard choice. For purity I'd lean into using physical type names as keys. Consistent with JSON encoding too. It is also consistent with logical types for which we don't have special handling (i.e. Duration).

However, this doesn't seem to buy us much but then the names in iceberg becomes less descriptive. I.e. from Iceberg POV I'd much rather look at a struct `{timestamp: [...], date: [...]}` than `{long: [...], int: [...]}`...

The downside is that it allows for some collisions. I.e. avro permits a logical type of timestamp with a record named timestamp but we wouldn't.

The cost looks like worthwhile to me.

Please add test examples for logical types.

oleiman replied:
two ambiguities that are a bit concerning:
- both int & long can map to "time" (logical types "time-millis" and "time-micros", respectively)
- both bytes & fixed can map to "decimal"

> avro permits a logical type of timestamp with a record named timestamp but we wouldn't

Yeah this is not great. There seems to be a rule against naming records w/ a primitive type name, but no rule against logical type names.

I'm leaning more towards the rationale in https://github.com/risingwavelabs/risingwave/issues/17616 so we can let the Avro library enforce label collisions in the record schema.

rockwotj reviewed (dismissed):
LGTM, this is way nicer. Just one question.

rockwotj commented on src/v/iceberg/conversion/tests/iceberg_avro_tests.cc:
```diff
@@ -1269,3 +1269,114 @@ INSTANTIATE_TEST_SUITE_P(
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "fixed", "name": "FixedDecimal", "size": 16, "logicalType": "decimal", "precision": 10, "scale": 2}, "long" ])J",
```
why aren't we using the `name` here? Shouldn't we?

oleiman replied:
oh, that was an accident. good catch

oleiman replied:
fixed

rockwotj replied:
Since these decision are hard to revert, can we make sure we don't miss any other cases?

oleiman replied:
Added a test case for symbolic name resolution for sanity. Other than that I was just grepping through `conversion/schema_avro.cc` for logical types. In that respect these cases should be exhaustive.
